### ISSUE TITLE
conversion to utc when offset is present is a bad idea

### ISSUE THREAD

**Issue by marshallpierce** [temporenc/temporenc#8]
I could be missing something here (it wouldn't be the [first time](https://github.com/wbolster/temporenc/pull/7)) but I'm not convinced that applying the specified offset to the date and time before encoding, aka storing the date and time in UTC, is a good idea. 

Applying a UTC offset to a given set of date and time fields necessarily involves applying a calendar system. By default that would probably be Gregorian (with the "leap year every 4, except on 100 year boundaries, but actually still do it every 400), but the year scale in Temporenc goes back to 0, so before October 1582 it would need to be the Julian calendar back to AD 8 (when the extra day in February every 4 years showed up), and before that it would be the previous version of the Julian calendar without the leap day. Then, of course, there is the consideration of other calendar systems. Even if an implementation got all that right, what about other locales? The Indian national calendar (which applies to fifth of the world's people) has different leap year and leap day rules (and different month lengths), so it would be totally different under their rules. The point is, it's difficult to imagine interop working well across the calendar system boundaries even if you just assume it's Gregorian. This is both because different platforms will have different time libraries that make different assumptions about what calendar systems need to be supported, and because users will vary in their knowledge of this complexity. Someone who knows about the switch to the Gregorian calendar would take it into account, and someone who doesn't could be surprised when they deserialize an offset value. 

Furthermore, once you enter into the domain of requiring application of a calendar system to handle offset time, that means that any situation that's potentially troublesome to a rich calendar system now applies to Temporenc. What about the case where someone stores an offset timestamp during a ["normal" leap second](https://hpiers.obspm.fr/eoppc/bul/bulc/bulletinc.49) and serializes that, which is then deserialized by another system (possibly even using the same Temporenc implementation) but with different system configuration and leap second rules, like Google's "smeared" leap seconds? The leap second is technically not expressible on such a system, so even though it's not a problem for the encoded Temporenc value, the postprocessing to apply the offset is error-prone, complex and highly situation dependent. There is no one right answer there, so users would need to take matters into their own hands based on what they're doing. This becomes much harder when the interactions with the calendar system are happening *within* their Temporenc implementation. How should one allow plugging in a policy to serialization/deserialization to allow customization, etc? Also, this means that deserializing a timestamp on two different machines with two different Olson databases, leap second configurations, etc could yield different moments in time, which seems bad.

The sorting behavior is also strange to my eyes (obviously, since I missed it when implementing this spec ;)). First, there's just the surprise factor of having two ISO 8601 timestamps, a format already crafted for lexical sorting, change sort position when encoded in Temporenc. This pair of ISO 8601 lexically sorted timestamps would get swapped, for instance:

- 2017-01-15T12:01:02+02:00
- 2017-01-15T13:01:02+06:00

I'm sure that there are situations where it's useful to store the original offset but sort based on the un-offset fields, but that is *not* the typical assumption, which, for better or for worse, is ISO 8601's rules. It means that Temporenc cannot be thought of as "ISO 8601, but more compact". Instead, there's this wrinkle of lexical sorting now being sensitive to offset. To people who are already storing things in UTC, it does them no benefit, and to people who are storing things in their original offset in ISO 8601, it's wrong, because now they have different sort behavior. Anecdotally, in practice I see systems who care about global ordering handling everything in UTC anyway, so I'm not sure how widely applicable the use case of this feature of the encoding is. I'm sure there ARE cases where it's useful; I've just never seen them, and thus I have a hard time believing that it's worth sacrificing ISO-8601 sort order consistency, the principle of least surprise, the ability to sort based on the offset fields, etc.

So, to wrap up, it seems to me that this one aspect brings a lot of complexity and potential for correctness and interop issues to support a use case that is at best of limited utility and at worst a deal-breaker for those looking to replace ISO-8601. My preference would be instruct users the following:

- If they want a global ordering of timestamps by the instant they represent on UTC, then they should serialize in UTC and byte-wise sort
- If they want a global ordering by their lexical sort order with offset fields, then they should leave them offset and byte-wise sort
- If they have serialized timestamps with offsets but they want them ordered by UTC, then they should deserialize them and use the time library and conversion rules of their choice to map them to UTC, then sort the resulting instants.

In other words, just like they do now with ISO 8601.

**Comment by wbolster** [comment]
i think you're right.

the sort property was the main reason for this choice, but in reality systems that both a) handle time zone info and b) care about sorting, will do the sane thing and use utc to get ordering right.

though i came up with temporenc over two years ago, i consider this format still young (it is not widely used, if at all) so i am open for making breaking changes.

funny to see how this side project of mine suddenly gained since interest from others. apparently i am not the only one obsessing about this topic. ;)

**Comment by marshallpierce** [comment]
Obsessing over time is one of my favorite hobbies! ;) I saw the link to this project in @jodastephen's tweet and immediately was glad someone else had done the fiddly work of finding a binary scheme that was better than Unix timestamp.

You could always go the route of having "profiles" or something like that (profile A for encoded-with-utc-fields, profile B for encoded-as-is), but that will probably lead to long term confusion, so if you're up for it I vote for ripping the bandage off all at once.

**Comment by wbolster** [comment]
ok just to make it clear: utc conversion is a bad idea. i will remove it.

i am not a fan of profiles since it would add another dimension of complexity. i prefer to keep things as simple as possible unless there is a good reason for complexity. for example: variable length bit shifting? yes! because compactness!

while making breaking changes, what else is not as awesome as it should be? now is the time to fix all teh things. please open bugs (or tweet me) for discussion. :)

**Comment by marshallpierce** [comment]
That all sounds great. I'm wrapping up my rust implementation soon; I'll let you know if I find anything else that seems sub-optimal.

**Comment by jodastephen** [comment]
I agree with removing the pre-convert to UTC requirement. The Java type `OffsetDateTime` should be the model here, with each field (year, month, day, hour, min, sec, offset) mapping straight into and out of the binary format.

If you are going to break compatibility with what you've stated for 2 years, I recommend calling it v2. No-one will care about v1 as its not widely used, but it does show care for compatibility, which people like.

**Comment by wbolster** [comment]
@marshallpierce @jodastephen any chance you can have a look at #9? anything that should be different apart from fixing the sample values?

**Comment by wbolster** [comment]
fyi i have adapted my python implementation to no longer do utc conversion, which involved dropping all date arithmetic code, and fixing the tests. i like it.

https://github.com/temporenc/temporenc-python/pull/21
